File: swarmmanagercontroller/src/main/java/co/uk/swarmbit/docker/api/secrets/SecretsApiImpl.java
package co.uk.swarmbit.docker.api.secrets;

import co.uk.swarmbit.docker.api.common.AbstractApiImpl;
import co.uk.swarmbit.docker.api.common.json.SecretCreateResponseJson;
import co.uk.swarmbit.docker.api.secrets.parameters.SecretCreateParameters;
import co.uk.swarmbit.docker.api.secrets.parameters.SecretUpdateParameters;
import co.uk.swarmbit.docker.api.secrets.parameters.SecretsListParameters;
import co.uk.swarmbit.rest.RestResponseType;
import co.uk.swarmbit.docker.api.common.json.SecretJson;
import org.springframework.stereotype.Component;

import java.util.List;

@Component
public class SecretsApiImpl extends AbstractApiImpl implements SecretsApi {

    private static final String SECRETS_PATH = "/secrets";

    private static final String CREATE_PATH = "/create";

    private static final String UPDATE_PATH = "/update";

    @Override
    public List<SecretJson> listSecrets(String swarmId, SecretsListParameters parameters) {
        return listObjects(SECRETS_PATH, swarmId, new RestResponseType<List<SecretJson>>() {}, parameters);
    }

    @Override
    public SecretJson inspectSecret(String swarmId, String id) {
        return inspectObject(SECRETS_PATH, swarmId, new RestResponseType<SecretJson>() {}, id);
    }

    @Override
    public SecretCreateResponseJson createSecret(String swarmId, SecretCreateParameters parameters) {
        return createObject(SECRETS_PATH + CREATE_PATH, swarmId, new RestResponseType<SecretCreateResponseJson>() {}, parameters);
    }

    @Override
    public void deleteSecret(String swarmId, String id) {
        deleteObject(SECRETS_PATH + "/" + id, swarmId, new RestResponseType<Void>() {});
    }

    @Override
    public void updateSecret(String swarmId, String id, SecretUpdateParameters parameters) {
        updateObject(SECRETS_PATH + "/" + id + UPDATE_PATH, swarmId, new RestResponseType<Void>() {}, parameters, parameters);
    }
}

File: swarmmanagercontroller-new/src/main/kotlin/com/swarmbit/docker/api/secrets/SecretsApiImpl.kt
package com.swarmbit.docker.api.secrets

import com.swarmbit.docker.api.common.AbstractApiImpl
import com.swarmbit.docker.api.common.client.DockerWebClient
import com.swarmbit.docker.api.common.rest.model.ResponseType
import com.swarmbit.docker.api.secrets.model.Secret
import com.swarmbit.docker.api.secrets.model.SecretCreateResponse
import com.swarmbit.docker.api.secrets.parameters.SecretsCreateParameters
import com.swarmbit.docker.api.secrets.parameters.SecretsListParameters
import com.swarmbit.docker.api.secrets.parameters.SecretsUpdateParameters
import io.micronaut.context.annotation.Context

@Context
class SecretsApiImpl(dockerWebClient: DockerWebClient): AbstractApiImpl(dockerWebClient), SecretsApi {

    companion object {
        private const val SECRETS_PATH = "/secrets"
        private const val CREATE_PATH = "/create"
        private const val UPDATE_PATH = "/update"
    }

    override fun listSecrets(swarmId: String, parameters: SecretsListParameters): List<Secret> {
        return listObjects(SECRETS_PATH, swarmId, object : ResponseType<List<Secret>>() {}, parameters)
    }

    override fun inspectSecret(swarmId: String, id: String): Secret {
        return inspectObject(SECRETS_PATH, swarmId, object : ResponseType<Secret>() {}, id)
                ?: throw IllegalArgumentException("No secret found for swarmId ($swarmId) and configId (${id})")
    }

    override fun createSecret(swarmId: String, parameters: SecretsCreateParameters): SecretCreateResponse {
        return createObject(SECRETS_PATH + CREATE_PATH, swarmId, object : ResponseType<SecretCreateResponse>() {}, parameters)!!
    }

    override fun deleteSecret(swarmId: String, id: String) {
        deleteObject("$SECRETS_PATH/$id", swarmId, object : ResponseType<Void>() {})
    }

    override fun updateSecret(swarmId: String, id: String, parameters: SecretsUpdateParameters) {
        updateObject("$SECRETS_PATH/$id$UPDATE_PATH", swarmId, object : ResponseType<Void>() {}, parameters, parameters)

    }

}